### DISCUSSION TITLE
Making a `config:set` change during a git push deploy?

### DISCUSSION THREAD

**Discussion by beporter** [dokku/dokku#4590] [Q&A]
I searched the docs a bunch but couldn't find the cross-section of these two things.

If I wanted to _change_ an existing ENV var _during_ a `git push`-based deploy to a herokuish buildpack, is there a way to accomplish this? Otherwise it seems like I need to have excellent timing to run the `dokku config:set appname VARNAME=value` command immediately after the `git push dokku-host` command completes.

What am I missing?

**Comment by beporter** [comment]
I've discovered the `--no-restart` option to pre-set the new value, but this still doesn't necessarily always work (in our case, we need the ENV var available at the next build-time but ALSO runtime.)

**Comment by nickhammond** [comment]
@beporter Can you provide some context around what you're attempting to do with the environment variable? With the `--no-restart` flag your currently running containers wouldn't pick up the config change until the container is restarted, basically ENV won't be refreshed until a new version of that container starts.

**Reply by beporter** [reply]
The big issue is that there's no way to "schedule" a Dokku config change as a part of a git-based deploy. It _requires_ a manual intervention on the server first, before the deploy occurs. 

If you have CI set up to trigger the `git push`, say when your test suite passes, making the manual config change at exactly the right time can be both inaccurate and error prone. Like database migrations, this is the kind of thing that you'd want to happen "with" the deploy automatically, and not leave up to humans.

**Reply by nickhammond** [reply]
I'm still not entirely sure what the specific use case is but it sounds like an external config store is going to be more useful for you. Something you can fetch a configuration value from at any point whether it's on boot or restart. How does the config value that you're talking about get calculated? "Scheduling" a config change after you've successfully released something opens the app up to a bit of uncertainty, you want the fully released artifact to run as-is.

**Comment by josegonzalez** [comment]
Your CI should set apply any settings (with no-restart) and then proceed with a deploy. Unfortunately, there isn't any way to apply env vars at deploy time, as I think the desired thing would be to only persist them on successful deploy, and that gets dicey due to the various ways in which a deploy can fail.

The other reason we don't support something like this is because secrets should not be included _inside_ of a repository in unencrypted form. Instead of allowing something that would be potentially insecure, we won't even support the notion of it. We could introduce some sort of decryption process, but I suspect that will be more hassle than it's worth. I could be convinced otherwise, but I'd rather not introduce a footgun if at all possible.

As an alternative, you could implement a plugin that injects the config during the `post-extract` hook. That plugin can read the env keys from like... the `app.json` file and apply them as desired. Here is a (hella untested, please don't abuse this) example hook written in python:

```python
#!/usr/bin/env python3
import base64
import json
import os
import subprocess
import sys

app = sys.argv[0]
tmp_work_dir = sys.argv[1]

appjson_path = os.path.join(tmp_work_dir, "app.json")
if not os.path.exists(appjson_path):
    sys.exit(0)

content = ""
with open(appjson_path) as f:
    content = f.read()

if content == "":
    sys.exit(0)

data = {}
try:
    data = json.loads(content)
except:
    print(" !     Error reading app.json", file=sys.stderr)
    sys.exit(1)

env = data.get("env", {})

if type(env) is not dict:
    print(" !     Env key of app.json is malformed, expected dictionary", file=sys.stderr)
    sys.exit(1)

print('-----> Applying environment variables from app.json')

command = ["dokku", "--app", app, "config:set", "--no-restart", "--encoded"]
for key, config in env.items():
    if type(config) is not dict:
        print(" !     Config for {0} in app.json is malformed, expected dictionary".format(key), file=sys.stderr)
        sys.exit(1)

    if "value" not in config:
        continue

    if "value" == "":
        continue

    value = base64.b64encode(bytes(config["value"], "utf-8"))
    command.append("{0}={1}", key, value.decode("utf-8"))

try:
    subprocess.run(command, check=True)
except subprocess.CalledProcessError:
    print(" !     Error setting environment variables from app.json", file=sys.stderr)
    sys.exit(1)

print("       Environment variables applied from app.json")
```

**Reply by josegonzalez** [reply]
@beporter heh I'm waffling on this. See #4498 for a proposal (if you'd like to contribute it in some way, that would be excellent!).
